lib/likelihood.c

```diff
@@ -0,0 +1,304 @@
+                ret = tsk_tree_get_parent(tree, parent->id, &node_id);
+                while (tsk_tree_get_parent(tree, parent->id, &node_id) == 0
+                        && tree->left_child[parent->id] 
+                        == tree->right_child[parent->id]) {
+                    *child = *parent;
+                    ret = tsk_treeseq_get_node(ts, node_id, parent);
+                    if (ret != 0) {
+                        goto out;
+                    }
+                    branch_length += parent->time - child->time;
+                }
+                ret = tsk_treeseq_get_node(ts, mut.node, child);
+                if (ret != 0) {
+                    goto out;
+                }
+                while (tree->left_child[child->id] >= 0 &&
+                        tree->left_child[child->id] 
+                        == tree->right_child[child->id]) {
+                    *parent = *child;
+                    node_id = tree->left_child[child->id];
+                    ret = tsk_treeseq_get_node(ts, node_id, child);
+                    if (ret != 0) {
+                        goto out;
+                    }
+                    branch_length += parent->time - child->time;
+                }
+                *lik += log(branch_length / total_material);
+            }
+        }
+    } else if (num_mutations > 0) {
+        *lik = -DBL_MAX;
+    } else {
+        *lik = 0;
+    }
+out:
+    if (tree != NULL) {
+        free(tree);
+    }
+    if (edge != NULL) {
+        free(edge);
+    }
+    if (parent != NULL) {
+        free(parent);
+    }
+    if (child != NULL) {
+        free(child);
+    }
+    return ret;
+}
+
+int
+msp_log_likelihood_arg(tsk_treeseq_t *ts, double rho, double *lik) {
+    int ret = 1;
+    int i;
+    int lineages = (int)tsk_treeseq_get_num_samples(ts);
+    tsk_id_t parent_id;
+    double sim_time = 0;
+    double material = lineages * tsk_treeseq_get_sequence_length(ts);
+    double material_in_children;
+    double material_in_parent;
+    double rate;
+    double gap;
+    int first_parent_edge[ts->tables->nodes.num_rows];
+    int last_parent_edge[ts->tables->nodes.num_rows];
+    tsk_edge_t *edge = NULL;
+    tsk_node_t *parent = NULL;
+    tsk_node_t *child = NULL;
+    edge = malloc(sizeof *edge);
+    if (edge == NULL) {
+        ret = MSP_ERR_NO_MEMORY;
+        goto out;
+    }
+    parent = malloc(sizeof *parent);
+    if (parent == NULL) {
+        ret = MSP_ERR_NO_MEMORY;
+        goto out;
+    }
+    child = malloc(sizeof *child);
+    if (child == NULL) {
+        ret = MSP_ERR_NO_MEMORY;
+        goto out;
+    }
+    for (i = 0; i < (int)ts->tables->nodes.num_rows; i++) {
+        first_parent_edge[i] = -1;
+        last_parent_edge[i] = -1;
+    }
+    for (i = 0; i < (int)ts->tables->edges.num_rows; i++) {
+        ret = tsk_treeseq_get_edge(ts, i, edge);
+        if (ret != 0) {
+            goto out;
+        }
+        if (first_parent_edge[edge->child] == -1) {
```

Comment:
Might be good to use TSK_NULL here rather than -1. Might be better to use ``tsk_id_t`` types rather than plain ints also.